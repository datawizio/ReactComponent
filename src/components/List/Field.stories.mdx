import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import List from "./index";

<Meta title="Components/Lists/List.Item.Field" component={List.Item.Field} />

# ListItemField

Компонент `List.Item.Field`.

## Basic

<Preview>
  <Story name="Basic">
    <List.Item.Field
      title="List item title"
      description="List item desctiption"
    />
  </Story>
</Preview>

## With value

<Preview>
  <Story name="With value">
    <List.Item.Field
      title="List item title"
      description="List item desctiption"
      value="Value"
    />
  </Story>
</Preview>

<Props of={List.Item.Field} />
