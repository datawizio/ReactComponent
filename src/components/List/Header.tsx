import React from "react";

import { List } from "antd";

import { ListItemHeaderProps } from "./types.d";

const ListItemHeader: React.FC<ListItemHeaderProps> = ({ title }) => {
  return <List.Item className="list-item-header">{title}</List.Item>;
};

export default ListItemHeader;
