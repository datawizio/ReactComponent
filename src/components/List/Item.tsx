import React from "react";
import { List } from "antd";
import { ListItemTypeProps } from "./types";
import ListItemField from "./Field";
import ListItemHeader from "./Header";

const Item: ListItemTypeProps = props => <List.Item {...props} />;

Item.Field = ListItemField;
Item.Header = ListItemHeader;
Item.Meta = List.Item.Meta;

export default Item;
