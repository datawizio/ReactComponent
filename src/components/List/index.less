@import "~antd/es/list/style/index";
@import "../../styles/variables";

.list-item-field {
  padding-left: @padding-xs;
  padding-right: @padding-xss;
  cursor: pointer;
  &:hover {
    background-color: @table-row-hover-bg;
  }
  .ant-list-item-meta-title {
    margin-bottom: 0;
  }
  .ant-list-item-meta-description {
    line-height: 1.3;
  }
}

.ant-list-item.list-item-header {
  font-size: 20px;
}
