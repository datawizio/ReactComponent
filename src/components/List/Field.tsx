import React from "react";

import List from "../List";
import { RightOutlined } from "@ant-design/icons";
import { ListItemFieldProps } from "./types.d";

const ListItemField: React.FC<ListItemFieldProps> = ({
  title,
  description,
  value,
  onClick
}) => {
  return (
    <List.Item onClick={onClick} className="list-item-field">
      <List.Item.Meta title={title} description={description} />
      <div>
        {value} <RightOutlined />
      </div>
    </List.Item>
  );
};

export default ListItemField;
