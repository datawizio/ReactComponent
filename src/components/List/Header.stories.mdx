import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import List from "./index";

<Meta title="Components/Lists/List.Item.Header" component={List.Item.Header} />

# ListItemField

Компонент `List.Item.Header`.

## Basic

<Preview>
  <Story name="Basic">
    <List.Item.Header title="List item header" />
  </Story>
</Preview>

<Props of={List.Item.Header} />
