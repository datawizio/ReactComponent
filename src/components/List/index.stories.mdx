import { Meta, Story, Preview, Props } from "@storybook/addon-docs/blocks";
import { useMemo, useState, useEffect, useCallback, useRef } from "react";
import server from "../../utils/data/server";
import faker from "faker";

import {
  withKnobs,
  text,
  boolean,
  number,
  select
} from "@storybook/addon-knobs";

import Avatar from "antd/lib/avatar";
import Divider from "antd/lib/divider";
import List, { _ListWithProps } from "./";
import InfiniteScroll from "../InfiniteScroll";

import "antd/lib/avatar/style/index";
import "antd/lib/divider/style/index";

<Meta title="Components/Lists/List" component={List} decorators={[withKnobs]} />

# List

Компонент `List`.

Базовый компонент: [**antd/list**](https://ant.design/components/list)

## Sandbox

<Preview>
  <Story name="Sandbox">
    {() => {
      const dataCount = number("data count", 10);
      const sizeType = select("size", ["small", "large", "default"], "default");
      //
      const loading = boolean("loading", false);
      const isBordered = boolean("bordered", true);
      const hasHeader = boolean("Show Header", false);
      const hasFooter = boolean("Show Footer", false);
      const withPagination = boolean("pagination", false);
      const hasExtraImage = boolean("Show extra image", true);
      //
      const headerText = hasHeader && text("header", "Header");
      const footerText = hasFooter && text("herder", "Footer");
      //
      const paginationProps = withPagination && {
        total: dataCount,
        pageSize: number("pageSize", 5),
        showSizeChanger: boolean("showSizeChanger", false)
      };
      //
      const data = useMemo(() => {
        return new Array(dataCount).fill(null).map(() => {
          return {
            avatar: faker.image.avatar(),
            fullName: faker.name.findName(),
            address: faker.address.streetAddress(),
            image: "https://picsum.photos/200" + "?" + faker.random.number()
          };
        });
      }, [dataCount]);
      //
      return (
        <List
          size={sizeType}
          loading={loading}
          dataSource={data}
          header={headerText}
          footer={footerText}
          bordered={isBordered}
          pagination={paginationProps}
          renderItem={item => {
            return (
              <List.Item
                extra={
                  hasExtraImage && (
                    <img height="100" width="150" src={item.image} />
                  )
                }
              >
                <List.Item.Meta
                  avatar={<Avatar src={item.avatar} />}
                  title={item.fullName}
                  description={item.address}
                />
              </List.Item>
            );
          }}
        />
      );
    }}
  </Story>
</Preview>

## Basic

<Preview>
  <Story name="Basic">
    {() => {
      const data = useMemo(() => {
        return new Array(10).fill(null).map(() => faker.name.findName());
      }, []);
      //
      return (
        <List
          dataSource={data}
          renderItem={item => {
            return <List.Item>{item}</List.Item>;
          }}
        />
      );
    }}
  </Story>
</Preview>

## Pagination

<Preview>
  <Story name="Pagination">
    {() => {
      const data = useMemo(() => {
        return new Array(100).fill(null).map(() => faker.name.findName());
      }, []);
      //
      return (
        <div>
          <Divider orientation="left">Basic pagination</Divider>
          <List
            pagination={{ defaultPageSize: 10 }}
            dataSource={data}
            renderItem={item => {
              return <List.Item>{item}</List.Item>;
            }}
          />
          <Divider orientation="left">Hidden size changer</Divider>
          <List
            pagination={{ showSizeChanger: false, pageSize: 10 }}
            dataSource={data}
            renderItem={item => {
              return <List.Item>{item}</List.Item>;
            }}
          />
        </div>
      );
    }}
  </Story>
</Preview>

## User List

<Preview>
  <Story name="User List">
    {() => {
      const data = useMemo(() => {
        return new Array(10).fill(null).map(() => {
          return {
            avatar: faker.image.avatar(),
            fullName: faker.name.findName(),
            address: faker.address.streetAddress()
          };
        });
      }, []);
      //
      return (
        <List
          dataSource={data}
          renderItem={item => {
            return (
              <List.Item>
                <List.Item.Meta
                  avatar={<Avatar src={item.avatar} />}
                  title={item.fullName}
                  description={item.address}
                />
              </List.Item>
            );
          }}
        />
      );
    }}
  </Story>
</Preview>

## Extra Images List

<Preview>
  <Story name="Extra Images List">
    {() => {
      const data = useMemo(() => {
        return new Array(10).fill(null).map(() => {
          return {
            content: faker.name.findName(),
            image: "https://picsum.photos/200" + "?" + faker.random.number()
          };
        });
      }, []);
      //
      return (
        <List
          dataSource={data}
          renderItem={item => {
            return (
              <List.Item
                extra={<img height="100" width="150" src={item.image} />}
              >
                {item.content}
              </List.Item>
            );
          }}
        />
      );
    }}
  </Story>
</Preview>

## Infinite scroll

<Preview>
  <Story name="Infinite scroll">
    {() => {
      const someData = useMemo(() => {
        return new Array(300).fill(null).map(() => {
          return {
            avatar: faker.image.avatar(),
            fullName: faker.name.findName(),
            address: faker.address.streetAddress()
          };
        });
      }, []);
      //
      const pageSize = 50;
      const [dataSource, setDataSource] = useState(someData.slice(0, pageSize));
      //
      return (
        <InfiniteScroll
          showLoader
          height={500}
          maxPage={someData.length / pageSize}
          onBottom={page => setDataSource(someData.slice(0, pageSize * page))}
        >
          <List
            dataSource={dataSource}
            loading={!dataSource.length}
            renderItem={item => {
              return (
                <List.Item>
                  <List.Item.Meta
                    title={item.fullName}
                    description={item.address}
                    avatar={<Avatar src={item.avatar} />}
                  />
                </List.Item>
              );
            }}
          />
        </InfiniteScroll>
      );
    }}
  </Story>
</Preview>

## Async scroll

<Preview>
  <Story name="Async scroll">
    {props => {
      const [dataSource, setDataSource] = useState([]);
      //
      const loadData = useCallback(
        async page => {
          const pageSize = 100;
          const response = await server.fetchUsers(pageSize, page);
          setDataSource(dataSource.concat(response));
        },
        [dataSource]
      );
      //
      useEffect(() => {
        loadData();
      }, []);
      //
      return (
        <InfiniteScroll height={500} onBottom={loadData} showLoader>
          <List
            dataSource={dataSource}
            loading={!dataSource.length}
            renderItem={item => {
              return (
                <List.Item>
                  <List.Item.Meta
                    title={item.fullName}
                    description={item.address}
                    avatar={<Avatar src={item.avatar} />}
                  />
                </List.Item>
              );
            }}
          />
        </InfiniteScroll>
      );
    }}
  </Story>
</Preview>

- [**antd/List.baseProps**](https://ant.design/components/list/#List)

<Props of={_ListWithProps} />
