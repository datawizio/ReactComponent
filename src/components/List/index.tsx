import * as React from "react";
import AntdList from "antd/lib/list";
import Item from "./Item";

import { FCList, ListProps } from "./types.d";

import "./index.less";

const List: FCList = props => {
  return <AntdList {...props} />;
};

List.defaultProps = {
  bordered: true
};

List.Item = Item;

export default List;
export const _ListWithProps: React.FC<ListProps> = {} as any;
