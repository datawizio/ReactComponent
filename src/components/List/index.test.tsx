import "jsdom-global/register";
import React from "react";
import { shallow } from "enzyme";

import List from "./index";

const mockProps = {
  bordered: false
};

const setUp = (props?) => shallow(<List {...props} />);

describe("List component", () => {
  let component;
  beforeEach(() => {
    component = setUp(mockProps);
  });

  it("List render  with correct props", () => {
    expect(component.prop("bordered")).toBeFalsy();
  });
});
