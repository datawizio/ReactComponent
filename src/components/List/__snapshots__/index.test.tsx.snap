// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component List rendered correctly 1`] = `
<List
  bordered={false}
/>
`;
